Comprehensive Refactoring and Optimization of Project for Enhanced Readability, Efficiency, and Robustness

RoachxD opened the pull request:
This Pull Request encompasses a series of improvements and refinements to the project's structure, functionality, and code quality, with the objective of enhancing readability, efficiency, and error handling.

The primary contributions include:

- Optimization of `setup.py`: Refactoring the file for increased efficiency and readability, including utilization of the `Path` object from `pathlib` for better file path handling, simplification of `RegEx` patterns, and employment of `f-string` for string formatting.
- Refactoring of extension scripts: Adhering to Python's naming conventions and optimizing various aspects of the code, such as list comprehensions and `f-string` usage, to improve consistency, readability, and efficiency.
- Improved formatting of `conf.py`: Enhancing the readability and consistency of the file by removing blank lines, refining formatting, and employing list indexing with `re.search`.
- Implementation of the `all` variable: Explicitly defining public members for import, promoting a clean API and preventing unintended imports of private or internal members.
- Refinement of `request.py`: Removing unnecessary imports, utilizing the params keyword argument for improved readability, and streamlining error handling through `response.raise_for_status()` and combined exception handling.
- Division of `ESteamCurrency` into two classes: Separating supported and unsupported currencies to facilitate better error handling and functionality, while acknowledging potential limitations in regards to IDEs autocompletion.
- Introduction of a new exception: Adding `SteamUnsupportedCurrency` exception to handle cases where an unsupported currency is encountered.
- Implementation of float parsing for all currencies: Incorporating the `re` package for accurate parsing of both supported and unsupported currencies, and optimizing various aspects of the code, such as type hints, list comprehensions, and conditional statements, for improved readability and efficiency.

These cumulative enhancements result in a more streamlined, robust, and maintainable codebase that adheres to best practices and provides a solid foundation for future development.

RoachxD committed:
Added `.gitignore` to the project!

RoachxD committed:
Optimized and refactored `setup.py`!
 - Imported `Path` from `pathlib` to handle file paths more effectively.
 - Removed `find_packages` import as it isn't being used within the setup.
 - Changed the way the file paths are constructed, making use of the `Path` object.
 - Simplified the Regex pattern formatting of `version` and used list indexing to get the result from `re.search`'s result.
 - Used `read_text()` method for reading the content of the `README.rst` file.
 - Replaced string concatenation with `f-string` for the `download_url` parameter in the `setup()` function.

 This commit includes multiple improvements to the setup script, making it more efficient and readable.

RoachxD committed:
Formatted and fixed `RSPEC-1192`
https://rules.sonarsource.com/python/RSPEC-1192

RoachxD committed:
Optimize and refactor `extensions` scripts

`extensions/builder.py`:
- Use list comprehension for indexcounts generation.
- Replace string concatenation with `f-string` in `handle_page` call for `genindex-single.html`.

`extensions/details.py`:
- Change class names to follow Python's naming convention (capitalized class names).
- Remove extra line breaks within function bodies.

`extensions/exception_hierarchy.py`:
- Rename `exception_hierarchy` class to follow Python's naming convention (capitalized class names)

This commit includes improvements to the extensions scripts, making them more efficient, readable, and consistent.

RoachxD committed:
Optimize and refactor `conf.py`
 - Used list indexing when getting the `version` using `re.search`.
 - Remove blank lines and improve overall formatting for better readability.

This commit includes improvements to `conf.py`, making it more readable and consistent.

RoachxD committed:
Add `all` variable to export public members from this module

 - Defined `all` variable to list public members for when this module will be imported.
 - Ensure only `AppID`, `ESteamCurrency`, `Market`, and `request` are available for importing.

This commit adds the `all` variable to the script, explicitly defining the public members of the module that should be available for import when using a wildcard import statement (e.g. `from steam_community_market import *`). This helps maintain a clean API and prevents unintended imports of private or internal members.

RoachxD committed:
Formatted and optimized `request.py`
 - Removed the `import json` statement, as it is not required. The `requests` library provides a `.json()` method for parsing JSON responses.
 - Changed the `requests.get(url, payload)` to `requests.get(url, params=payload)` to use the params keyword argument. This makes the code more readable and ensures that the payload is sent as query parameters.
 - Added a call to `response.raise_for_status()` which raises a `requests.exceptions.RequestException` if the response contains an HTTP error status code. This simplifies the error handling process.
 - Combined the `requests.exceptions.RequestException` and `ValueError` exceptions into a single except block. Since the error handling for both exceptions is the same, it's more concise to catch both in a single block.

RoachxD committed:
Divided `ESteamCurrency` in two different classes
 - Added all currencies inside the enum, even legacy ones (that are no longer being supported by Steam) so we can later error handle them.
  - Divided `ESteamCurrency` into `ESteamSupportedCurrency` and `ESteamUnsupportedCurrency` using a 'hacky' way.

The only issue that arrises from doing this is that IDEs will no longer autocomplete `ESteamCurrency`'s properties.

We can/should create an issue related to this.

RoachxD committed:
Created `SteamUnsupportedCurrency` exception

RoachxD committed:
Implemented `float` parsing for all currencies, including the "old" `UNSUPPORTED_CURRENCY` ones
 - The `sys` import has been removed as it was not used in the code.
 - Removed `UNSUPPORTED_CURRENCY` as it is no longer needed.
  - Fixed different typos inside the `docstrings`.
 - Type hints have been added to function parameters and return types using the typing module, making it easier to understand the expected input and output types.
 - Changed the `__init__` method to raise a `SteamUnsupportedCurrency` exception in case the `currency` is not supported (instance of `ESteamUnsupportedCurrency`), and to use a newly implemented static method, called `_supported_currency`, in order to set the `currency` value non-dependant on the input format to `ESteamSupportedCurrency`.
 - Created the `_app_id_value` static method in order to set the `app_id`  value non-dependant on the input format, in both `get_overview` and `get_overviews`.
 - Simplified the conditional return statement inside the `get_overview`, `get_prices` and `get_volume` methods.
 - Used list comprehensions to make the code more concise and readable in both `get_overviews` and `get_overviews_from_dict`.
 - Used the `is` keyword when comparing an object to `None`, as `None` is a singleton object.
 - The code behind `get_lowest_price` and `get_median_price` has been implemented under `get_price`, which is how the two reach it as well.

The cherry on top is the fix that allows us to return floats from the previously unsupported currencies: `price_to_float` now correctly parses every currency correctly with the use of the `re` (RegEx) package.

RoachxD committed:
Upgraded to `1.2.4`!

Gobot1234 commented on setup.py:
```diff
@@ -20,13 +23,13 @@
     ],
     install_requires=["requests"],
-    python_requires='>=3.6',
+    python_requires=">=3.6",
```
This is no longer correct due to the usage of the new union syntax

RoachxD replied:
Correct, modified to `>=3.10`.

Gobot1234 commented on steam_community_market/market.py:
```diff
@@ -46,222 +58,219 @@ def get_overview(self, name: str, app_id) -> dict:
+        :rtype: :class:`int`
 
-        .. versionadded:: 1.2.0
+        .. versionadded: 1.2.4
```
This semi-colon shouldn't be removed

RoachxD replied:
Fixed, many thanks for pointing it out!

RoachxD committed:
After-review fixes!

RoachxD committed:
Added `SteamUnsupportedCurrency` to the `all` var

RoachxD committed:
Better formatted documentation and..
 More, documentation:
  - Added `description` and `note` to the `Currency` enums.
 - Added `description` to the `AppID` enum.
 - Fully documented the newly implemented `SteamUnsupportedCurrency` exception, the `Market` class and the `request` method.

RoachxD committed:
Added `REQUEST_HEADERS`
 - Generic headers have been added for when interacting with the Steam Community Market API.

RoachxD committed:
Added more app ids to the `AppID` enum
 - Documented their real name.
